src/server/scene/application_session.cpp

```diff
@@ -44,6 +46,41 @@ namespace mg = mir::graphics;
 namespace mev = mir::events;
 namespace mc = mir::compositor;
 
+class ms::ApplicationSession::SurfaceObserver : public NullSurfaceObserver
+{
+public:
+    explicit SurfaceObserver(ApplicationSession* session) : session{session} {}
+
+    /// Overrides from scene::SurfaceObserver
+    ///@{
+    void attrib_changed(Surface const* surf, MirWindowAttrib attrib, int /*value*/) override
+    {
+        switch (attrib)
+        {
+        case mir_window_attrib_state:
+        case mir_window_attrib_visibility:
+            session->track_overlapping_outputs(surf);
+            break;
+        default:;
+        }
+    }
+    void content_resized_to(Surface const* surf, geometry::Size const& /*content_size*/) override
+    {
+        session->track_overlapping_outputs(surf);
+    }
+    void moved_to(Surface const* surf, geometry::Point const& /*top_left*/) override
+    {
+        session->track_overlapping_outputs(surf);
+    }
+    void placed_relative(Surface const* surf, geometry::Rectangle const& /*placement*/) override
+    {
+        session->track_overlapping_outputs(surf);
+    }
+    ///@}
+private:
+    ApplicationSession* session;
+};
+
```

Comment:
We are observing the surface so that we can tell the surface that something changed?

That's weird - the surface already knows something changed, it just needs to decide what notifications are needed.

Reply:
Fixed with [17dd081](https://github.com/canonical/mir/pull/3244/commits/17dd081983813c68e316aac08f7d93810790d093).